File: Caronae/Ride/CreateRideViewController.m
@import ActionSheetPicker_3_0;
@import SVProgressHUD;

#import "CaronaeAlertController.h"
#import "CreateRideViewController.h"
#import "NSDate+nextHour.h"
#import "ZoneSelectionViewController.h"
#import "Caronae-Swift.h"

@interface CreateRideViewController () <UITextViewDelegate, ZoneSelectionDelegate>

@property (nonatomic) CGFloat routinePatternHeightOriginal;
@property (nonatomic) NSString *notesPlaceholder;
@property (nonatomic) UIColor *notesTextColor;
@property (nonatomic) NSDateFormatter *arrivalDateLabelFormatter;
@property (nonatomic) NSArray *hubs;
@end

@implementation CreateRideViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self checkIfUserHasCar];
    
    self.hubs = [CaronaeConstants defaults].centers;
    self.selectedHub = self.hubs.firstObject;
    
    self.rideDate = [NSDate nextHour];
    self.weekDays = [NSMutableArray arrayWithCapacity:7];
    self.routineDurationMonths = 2;
    
    self.arrivalDateLabelFormatter = [[NSDateFormatter alloc] init];
    self.arrivalDateLabelFormatter.dateFormat = @"dd/MM/yyyy HH:mm";
    [self.arrivalTimeButton setTitle:[self.arrivalDateLabelFormatter stringFromDate:self.rideDate] forState:UIControlStateNormal];
    
    self.segmentedControl.layer.cornerRadius = 8.0;
    self.segmentedControl.layer.borderColor = [UIColor colorWithWhite:0.690 alpha:1.000].CGColor;
    self.segmentedControl.layer.borderWidth = 2.0f;
    self.segmentedControl.layer.masksToBounds = YES;
    
    self.notes.layer.cornerRadius = 8.0;
    self.notes.layer.borderColor = [UIColor colorWithWhite:0.902 alpha:1.000].CGColor;
    self.notes.layer.borderWidth = 2.0f;
    self.notes.textContainerInset = UIEdgeInsetsMake(10, 5, 5, 5);
    self.notes.delegate = self;
    self.notesPlaceholder = self.notes.text;
    self.notesTextColor = self.notes.textColor;
    
    self.slotsLabel.text = [NSString stringWithFormat:@"%.f", self.slotsStepper.value];
    
    // Dismiss keyboard when tapping the view
    [self.view addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self.view action:@selector(endEditing:)]];
    
    NSDictionary *lastRideLocation = [[NSUserDefaults standardUserDefaults] objectForKey:@"lastOfferedRideLocation"];
    if (lastRideLocation) {
        self.zone = lastRideLocation[@"zone"];
        self.neighborhood = lastRideLocation[@"neighborhood"];
        if (self.neighborhood) {
            [self.neighborhoodButton setTitle:self.neighborhood forState:UIControlStateNormal];
        }
        if (lastRideLocation[@"place"]) self.reference.text = lastRideLocation[@"place"];
        if (lastRideLocation[@"route"]) self.route.text = lastRideLocation[@"route"];
        if (lastRideLocation[@"hubGoing"]) {
            self.selectedHub = lastRideLocation[@"hubGoing"];
            [self.center setTitle:self.selectedHub forState:UIControlStateNormal];
        }
        if (lastRideLocation[@"description"]) self.notes.text = lastRideLocation[@"description"];
    }
}

- (void)checkIfUserHasCar {
    if (!UserService.instance.user.carOwner) {
        [CaronaeAlertController presentOkAlertWithTitle:@"Você possui carro?" message:@"Parece que você marcou no seu perfil que não possui um carro.\n\nPara criar uma carona, preencha os dados do seu carro no seu perfil." handler:^{
            [self goBack:nil];
        }];
    }
}

- (IBAction)goBack:(id)sender {
    [SVProgressHUD dismiss];
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (Ride *)generateRideFromView {
    NSString *description = [self.notes.text isEqualToString:_notesPlaceholder] ? @"" : self.notes.text;
    BOOL going = (self.segmentedControl.selectedSegmentIndex == 0);
    
    Ride *ride = [[Ride alloc] init];
    ride.region = self.zone;
    ride.neighborhood = self.neighborhood;
    ride.place = self.reference.text;
    ride.route = self.route.text;
    ride.hub = self.selectedHub;
    ride.notes = description;
    ride.going = going;
    ride.date = self.rideDate;
    ride.slots = self.slotsStepper.value;
    
    // Routine fields
    if (self.routineSwitch.on) {
        NSString *weekDaysString = [[self.weekDays sortedArrayUsingSelector:@selector(localizedCaseInsensitiveCompare:)] componentsJoinedByString:@","];
        ride.weekDays = weekDaysString;
        
        // Calculate final date for event based on the selected duration
        NSDateComponents *dateComponents = [[NSDateComponents alloc] init];
        dateComponents.month = self.routineDurationMonths;
        NSDate *repeatsUntilDate = [[NSCalendar currentCalendar] dateByAddingComponents:dateComponents toDate:self.rideDate options:0];
        ride.repeatsUntil = repeatsUntilDate;
    }
    
    return ride;
}

- (void)savePresetLocationZone:(NSString *)zone neighborhood:(NSString *)neighborhood place:(NSString *)place route:(NSString *)route hub:(NSString *)hub description:(NSString *)description going:(BOOL)going {
    NSDictionary *lastRidePresets = [[NSUserDefaults standardUserDefaults] objectForKey:@"lastOfferedRideLocation"];
    NSMutableDictionary *newPresets = [NSDictionaryOfVariableBindings(zone, neighborhood, place, route, description) mutableCopy];

    if (going) {
        newPresets[@"hubGoing"] = hub;
        if (lastRidePresets[@"hubReturning"]) {
            newPresets[@"hubReturning"] = lastRidePresets[@"hubReturning"];
        }
    } else {
        newPresets[@"hubReturning"] = hub;
        if (lastRidePresets[@"hubGoing"]) {
            newPresets[@"hubGoing"] = lastRidePresets[@"hubGoing"];
        }

    }
    
    [[NSUserDefaults standardUserDefaults] setObject:newPresets forKey:@"lastOfferedRideLocation"];
}

- (void)createRide:(Ride *)ride {
    [SVProgressHUD show];
    self.createRideButton.enabled = NO;

    [RideService.instance createRide:ride success:^(NSArray<Ride *> * _Nonnull createdRides) {
        [SVProgressHUD dismiss];
        [self dismissViewControllerAnimated:YES completion:nil];
    } error:^(NSError * _Nullable error) {
        [SVProgressHUD dismiss];
        self.createRideButton.enabled = YES;

        NSLog(@"Error creating ride: %@", error.localizedDescription);

        [CaronaeAlertController presentOkAlertWithTitle:@"Não foi possível criar a carona." message:error.localizedDescription];
    }];
}

- (IBAction)didTapCreateButton:(id)sender {
    // Check if the user selected the location and hub
    if (!self.zone || !self.neighborhood || !self.selectedHub) {
        [CaronaeAlertController presentOkAlertWithTitle:@"Dados incompletos" message:@"Ops! Parece que você esqueceu de preencher o local da sua carona."];
        return;
    }
    
    Ride *ride = [self generateRideFromView];
    [self savePresetLocationZone:ride.region neighborhood:ride.neighborhood place:ride.place route:ride.route hub:ride.hub description:ride.notes going:ride.going];
    
    // Check if the user has selected the routine details
    if (ride.repeatsUntil && ride.weekDays.length == 0) {
        [CaronaeAlertController presentOkAlertWithTitle:@"Dados incompletos" message:@"Ops! Parece que você esqueceu de marcar os dias da rotina."];
        return;
    }
    
    [self createRide:ride];
}

- (IBAction)slotsStepperChanged:(UIStepper *)sender {
    self.slotsLabel.text = [NSString stringWithFormat:@"%.f", sender.value];
}


#pragma mark - Routine selection buttons

- (IBAction)routineSwitchChanged:(UISwitch *)sender {
    [self.view endEditing:YES];
    if (sender.on) {
        [self.view layoutIfNeeded];
        _routinePatternHeight.constant = _routinePatternHeightOriginal;
        [UIView animateWithDuration:0.5 animations:^{
            [self.view layoutIfNeeded];
            self.routinePatternView.alpha = 1.0f;
        }];
    }
    else {
        [self.view layoutIfNeeded];
        _routinePatternHeightOriginal = _routinePatternHeight.constant;
        _routinePatternHeight.constant = 0;
        [UIView animateWithDuration:0.5 animations:^{
            [self.view layoutIfNeeded];
            self.routinePatternView.alpha = 0.0f;
        }];
    }
}

- (IBAction)routineMondayButtonTapped:(UIButton *)sender {
    sender.selected = !sender.selected;
    if (sender.selected) {
        [self.weekDays addObject:@"1"];
    }
    else {
        [self.weekDays removeObject:@"1"];
    }
}

- (IBAction)routineTuesdayButtonTapped:(UIButton *)sender {
    sender.selected = !sender.selected;
    if (sender.selected) {
        [self.weekDays addObject:@"2"];
    }
    else {
        [self.weekDays removeObject:@"2"];
    }
}

- (IBAction)routineWednesdayButtonTapped:(UIButton *)sender {
    sender.selected = !sender.selected;
    if (sender.selected) {
        [self.weekDays addObject:@"3"];
    }
    else {
        [self.weekDays removeObject:@"3"];
    }
}

- (IBAction)routineThursdayButtonTapped:(UIButton *)sender {
    sender.selected = !sender.selected;
    if (sender.selected) {
        [self.weekDays addObject:@"4"];
    }
    else {
        [self.weekDays removeObject:@"4"];
    }
}

- (IBAction)routineFridayButtonTapped:(UIButton *)sender {
    sender.selected = !sender.selected;
    if (sender.selected) {
        [self.weekDays addObject:@"5"];
    }
    else {
        [self.weekDays removeObject:@"5"];
    }
}

- (IBAction)routineSaturdayButtonTapped:(UIButton *)sender {
    sender.selected = !sender.selected;
    if (sender.selected) {
        [self.weekDays addObject:@"6"];
    }
    else {
        [self.weekDays removeObject:@"6"];
    }
}

- (IBAction)routineSundayButtonTapped:(UIButton *)sender {
    sender.selected = !sender.selected;
    if (sender.selected) {
        [self.weekDays addObject:@"7"];
    }
    else {
        [self.weekDays removeObject:@"7"];
    }
}

- (IBAction)routineDurationButtonTapped:(UIButton *)sender {
    sender.selected = YES;
    if (sender == self.routineDuration2MonthsButton) {
        self.routineDurationMonths = 2;
        self.routineDuration3MonthsButton.selected = NO;
        self.routineDuration4MonthsButton.selected = NO;
    }
    else if (sender == self.routineDuration3MonthsButton) {
        self.routineDurationMonths = 3;
        self.routineDuration2MonthsButton.selected = NO;
        self.routineDuration4MonthsButton.selected = NO;
    }
    else if (sender == self.routineDuration4MonthsButton) {
        self.routineDurationMonths = 4;
        self.routineDuration2MonthsButton.selected = NO;
        self.routineDuration3MonthsButton.selected = NO;
    }
}


#pragma mark - Other actions

- (IBAction)selectDateTapped:(id)sender {
    [self.view endEditing:YES];
    ActionSheetDatePicker *datePicker = [[ActionSheetDatePicker alloc] initWithTitle: (self.segmentedControl.selectedSegmentIndex == 0)? @"Chegada ao destino" : @"Saída da UFRJ" datePickerMode:UIDatePickerModeDateAndTime selectedDate:self.rideDate target:self action:@selector(timeWasSelected:element:) origin:sender];
    ((UIDatePicker *)datePicker).minimumDate = [NSDate currentHour];
    [datePicker showActionSheetPicker];
}

- (void)timeWasSelected:(NSDate *)selectedTime element:(id)element {
    self.rideDate = selectedTime;
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"dd/MM/yyyy HH:mm"];
    [self.arrivalTimeButton setTitle:[dateFormatter stringFromDate:selectedTime] forState:UIControlStateNormal];
}

- (IBAction)directionChanged:(UISegmentedControl *)sender {
    [self.view endEditing:YES];
    NSDictionary *lastRideLocation = [[NSUserDefaults standardUserDefaults] objectForKey:@"lastOfferedRideLocation"];
    if (sender.selectedSegmentIndex == 0) {
        self.hubs = [CaronaeConstants defaults].centers;
        if (lastRideLocation[@"hubGoing"]) {
            self.selectedHub = lastRideLocation[@"hubGoing"];
            [self.center setTitle:self.selectedHub forState:UIControlStateNormal];
        } else {
            self.selectedHub = self.hubs.firstObject;
            [self.center setTitle:self.selectedHub forState:UIControlStateNormal];
        }
    }
    else {
        self.hubs = [CaronaeConstants defaults].hubs;
        if (lastRideLocation[@"hubReturning"]) {
            self.selectedHub = lastRideLocation[@"hubReturning"];
            [self.center setTitle:self.selectedHub forState:UIControlStateNormal];
        } else {
            self.selectedHub = self.hubs.firstObject;
            [self.center setTitle:self.selectedHub forState:UIControlStateNormal];
        }
    }
}

- (IBAction)selectCenterTapped:(id)sender {
    [self.view endEditing:YES];
    
    NSUInteger selectedIndex = [self.hubs indexOfObject:self.selectedHub];
    if (selectedIndex == NSNotFound) {
        selectedIndex = 0;
    }
    
    [ActionSheetStringPicker showPickerWithTitle:@"Selecione um centro"
                                            rows:self.hubs
                                initialSelection:selectedIndex
                                       doneBlock:^(ActionSheetStringPicker *picker, NSInteger selectedIndex, id selectedValue) {
                                           self.selectedHub = selectedValue;
                                           [self.center setTitle:selectedValue forState:UIControlStateNormal];
                                       }
                                     cancelBlock:nil origin:sender];
}

- (void)hasSelectedNeighborhood:(NSString *)neighborhood inZone:(NSString *)zone {
    self.zone = zone;
    self.neighborhood = neighborhood;
    [self.neighborhoodButton setTitle:self.neighborhood forState:UIControlStateNormal];
}


#pragma mark - UITextView delegate

- (void)textViewDidBeginEditing:(UITextView *)textView {
    if ([textView.text isEqualToString:_notesPlaceholder]) {
        textView.text = @"";
        textView.textColor = _notesTextColor;
    }
    [textView becomeFirstResponder];
}

- (void)textViewDidEndEditing:(UITextView *)textView {
    if ([textView.text isEqualToString:@""]) {
        textView.text = _notesPlaceholder;
        textView.textColor = [UIColor lightGrayColor];
    }
    [textView resignFirstResponder];
}


#pragma mark - Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"ViewZones"]) {
        ZoneSelectionViewController *vc = segue.destinationViewController;
        vc.type = ZoneSelectionZone;
        vc.delegate = self;
    }
}

@end

File: Caronae/Ride/CreateRideViewController.swift
import UIKit
import SVProgressHUD
import ActionSheetPicker_3_0

class CreateRideViewController: UIViewController, NeighborhoodSelectionDelegate, HubSelectionDelegate {

    @IBOutlet weak var neighborhoodButton: UIButton!
    @IBOutlet weak var reference: UITextField!
    @IBOutlet weak var route: UITextField!
    @IBOutlet weak var centerButton: UIButton!
    @IBOutlet weak var routineSwitch: UISwitch!
    @IBOutlet weak var slotsStepper: UIStepper!
    @IBOutlet weak var slotsLabel: UILabel!
    @IBOutlet weak var notes: UITextView!
    @IBOutlet weak var segmentedControl: UISegmentedControl!
    @IBOutlet weak var createRideButton: UIButton!
    
    @IBOutlet weak var routinePatternView: UIView!
    @IBOutlet weak var routinePatternHeight: NSLayoutConstraint!
    
    @IBOutlet weak var routineDuration2MonthsButton: UIButton!
    @IBOutlet weak var routineDuration3MonthsButton: UIButton!
    @IBOutlet weak var routineDuration4MonthsButton: UIButton!
    @IBOutlet weak var dateButton: UIButton!
    
    var routinePatternHeightOriginal: CGFloat?
    var notesPlaceholder: String?
    var notesTextColor: UIColor?
    
    var weekDays = [String]()
    var routineDurationMonths: Int?
    var selectedDate: Date? {
        didSet {
            let dateString = dateFormatter.string(from: selectedDate!)
            dateButton.setTitle(dateString, for: .normal)
        }
    }
    
    let userDefaults = UserDefaults.standard
    let dateFormatter = DateFormatter()
    
    var selectedNeighborhood = String() {
        didSet {
            let selection = selectedNeighborhood.isEmpty ? "Bairro" : selectedNeighborhood
            self.neighborhoodButton.setTitle(selection, for: .normal)
        }
    }

    var selectedHub = String() {
        didSet {
            let selection = selectedHub.isEmpty ? "Centro Universitário" : selectedHub
            self.centerButton.setTitle(selection, for: .normal)
        }
    }
    
    var selectedZone = String()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        checkIfUserHasCar()
        
        dateFormatter.dateFormat = "dd/MM/yyyy HH:mm"
        selectedDate = max(Date.nextHour, Date.currentTimePlus(minutes: 5))
        routineDurationMonths = 2
        
        segmentedControl.layer.cornerRadius = 8.0
        segmentedControl.layer.borderColor = UIColor(white: 0.690, alpha: 1.0).cgColor
        segmentedControl.layer.borderWidth = 2.0
        segmentedControl.layer.masksToBounds = true
        
        // Configure direction titles according to institution
        segmentedControl.setTitle(PlaceService.Institution.goingLabel, forSegmentAt: 0)
        segmentedControl.setTitle(PlaceService.Institution.leavingLabel, forSegmentAt: 1)
        
        notes.layer.cornerRadius = 8.0
        notes.layer.borderColor = UIColor(white: 0.902, alpha: 1.0).cgColor
        notes.layer.borderWidth = 2.0
        notes.textContainerInset = UIEdgeInsetsMake(10, 5, 5, 5)
        notes.delegate = self
        notesPlaceholder = notes.text
        notesTextColor = notes.textColor
        
        if let lastOfferedRide      = userDefaults.dictionary(forKey: CaronaePreferenceLastOfferedRide.key) {
            if let lastZone         = lastOfferedRide[CaronaePreferenceLastOfferedRide.zone] as? String,
               let lastNeighborhood = lastOfferedRide[CaronaePreferenceLastOfferedRide.neighborhood] as? String,
               let lastPlace        = lastOfferedRide[CaronaePreferenceLastOfferedRide.place] as? String,
               let lastRoute        = lastOfferedRide[CaronaePreferenceLastOfferedRide.route] as? String,
               let lastSlots        = lastOfferedRide[CaronaePreferenceLastOfferedRide.slots] as? Double,
               let lastDescription  = lastOfferedRide[CaronaePreferenceLastOfferedRide.description] as? String {
                selectedZone = lastZone
                selectedNeighborhood = lastNeighborhood
                reference.text = lastPlace
                route.text = lastRoute
                slotsStepper.value = lastSlots
                if !lastDescription.isEmpty {
                    notes.text = lastDescription
                    notes.textColor = .black
                }
            }
            
            if let lastHubGoing = lastOfferedRide[CaronaePreferenceLastOfferedRide.hubGoing] as? String {
                selectedHub = lastHubGoing
            }
        }
        
        slotsLabel.text = String(format: "%.f", slotsStepper.value)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        self.view.endEditing(true)
    }
    
    func checkIfUserHasCar() {
        if let user = UserService.instance.user, !user.carOwner {
            CaronaeAlertController.presentOkAlert(withTitle: "Você possui carro?", message: "Parece que você marcou no seu perfil que não possui um carro.Para criar uma carona, preencha os dados do seu carro no seu perfil.", handler: {
                self.didTapCancelButton(self)
            })
        }
    }
    
    func generateRideFromView() -> Ride {
        let description = notes.text == notesPlaceholder ? "" : notes.text
        let going = self.segmentedControl.selectedSegmentIndex == 0
        
        let ride = Ride()
        ride.region = selectedZone
        ride.neighborhood = selectedNeighborhood
        ride.place = reference.text
        ride.route = route.text
        ride.hub = selectedHub
        ride.notes = description;
        ride.going = going;
        ride.date = selectedDate
        ride.slots = Int(slotsStepper.value)
        
        // Routine fields
        if routineSwitch.isOn {
            let weekDaysString = weekDays.sorted().joined(separator: ",")
            ride.weekDays = weekDaysString
            
            // Calculate final date for event based on the selected duration
            var dateComponents = DateComponents()
            dateComponents.month = routineDurationMonths
            let repeatsUntilDate = Calendar.current.date(byAdding: dateComponents, to: ride.date)
            ride.repeatsUntil = repeatsUntilDate;
        }
        
        return ride;
    }
    
    func savePreset(ride: Ride) {
        let lastRidePresets = userDefaults.dictionary(forKey: CaronaePreferenceLastOfferedRide.key)
        
        var newPresets: [String : Any] = [CaronaePreferenceLastOfferedRide.zone         : ride.region,
                                          CaronaePreferenceLastOfferedRide.neighborhood : ride.neighborhood,
                                          CaronaePreferenceLastOfferedRide.place        : ride.place,
                                          CaronaePreferenceLastOfferedRide.route        : ride.route,
                                          CaronaePreferenceLastOfferedRide.slots        : ride.slots,
                                          CaronaePreferenceLastOfferedRide.description  : ride.notes]
        
        if ride.going {
            newPresets[CaronaePreferenceLastOfferedRide.hubGoing] = ride.hub
            if let lastHubReturning = lastRidePresets?[CaronaePreferenceLastOfferedRide.hubReturning] {
                newPresets[CaronaePreferenceLastOfferedRide.hubReturning] = lastHubReturning
            }
        } else {
            newPresets[CaronaePreferenceLastOfferedRide.hubReturning] = ride.hub
            if let lastHubGoing = lastRidePresets?[CaronaePreferenceLastOfferedRide.hubGoing] {
                newPresets[CaronaePreferenceLastOfferedRide.hubGoing] = lastHubGoing
            }
        }
        
        userDefaults.set(newPresets, forKey: CaronaePreferenceLastOfferedRide.key)
    }
    
    func createRide(ride: Ride) {
        SVProgressHUD.show()
        createRideButton.isEnabled = false
        
        savePreset(ride: ride)
        
        RideService.instance.createRide(ride, success: {
            SVProgressHUD.dismiss()
            lastAllRidesUpdate = Date.distantPast
            self.dismiss(animated: true, completion: nil)
        }, error: { error in
            SVProgressHUD.dismiss()
            self.createRideButton.isEnabled = true
            
            NSLog("Error creating ride: %@", error.localizedDescription)
            
            CaronaeAlertController.presentOkAlert(withTitle: "Não foi possível criar a carona.", message: error.localizedDescription)
        })
    }
    
    @IBAction func didTapCreateButton(_ sender: Any) {
        self.view.endEditing(true)
        
        // Check if the user selected the location and hub
        if selectedZone.isEmpty || selectedNeighborhood.isEmpty || selectedHub.isEmpty {
            CaronaeAlertController.presentOkAlert(withTitle: "Dados incompletos", message: "Ops! Parece que você esqueceu de preencher o local da sua carona.")
            return
        }
        
        // Check if the user has selected the routine details
        if routineSwitch.isOn && weekDays.isEmpty {
            CaronaeAlertController.presentOkAlert(withTitle: "Dados incompletos", message: "Ops! Parece que você esqueceu de marcar os dias da rotina.")
            return
        }
        
        SVProgressHUD.show()
        createRideButton.isEnabled = false
        let ride = generateRideFromView()
        
        RideService.instance.validateRideDate(ride: ride, success: { isValid, status in
            SVProgressHUD.dismiss()
            if isValid {
                self.createRide(ride: ride)
            } else {
                switch status {
                case "duplicate":
                    CaronaeAlertController.presentOkAlert(withTitle: "Você já ofereceu uma carona muito parecida com essa", message: "Você pode verificar as suas caronas na seção 'Minhas' do aplicativo.", handler: {
                        self.createRideButton.isEnabled = true
                    })
                default:
                    let alert = CaronaeAlertController(title: "Parece que você já ofereceu uma carona para este dia", message: "Você pode cancelar e verificar as suas caronas ou continuar e criar a carona mesmo assim.", preferredStyle: .alert)
                    alert?.addAction(SDCAlertAction(title: "Cancelar", style: .cancel, handler: { _ in
                        self.createRideButton.isEnabled = true
                    }))
                    alert?.addAction(SDCAlertAction(title: "Criar", style: .recommended, handler: { _ in
                        self.createRide(ride: ride) }))
                    alert?.present(completion: nil)
                }
            }
            
        }, error: { error in
            SVProgressHUD.dismiss()
            self.createRideButton.isEnabled = true
            
            CaronaeAlertController.presentOkAlert(withTitle: "Não foi possível validar sua carona.", message: String(format: "Houve um erro de comunicação com nosso servidor. Por favor, tente novamente. (%@)", error!.localizedDescription))
        })
    }
    
    
    // MARK: IBActions
    
    @IBAction func didTapCancelButton(_ sender: Any) {
        SVProgressHUD.dismiss()
        self.dismiss(animated: true, completion: nil)
    }
    
    @IBAction func slotsStepperChanged(_ sender: UIStepper) {
        self.view.endEditing(true)
        slotsLabel.text = String(format: "%.f", sender.value)
    }
    
    @IBAction func routineSwitchChanged(_ sender: UISwitch) {
        view.endEditing(true)
        if sender.isOn {
            routinePatternHeight.constant = routinePatternHeightOriginal!
            UIView.animate(withDuration: 0.5, animations: {
                self.view.layoutIfNeeded()
                self.routinePatternView.alpha = 1.0
            })
        } else {
            routinePatternHeightOriginal = routinePatternHeight.constant
            routinePatternHeight.constant = 0
            UIView.animate(withDuration: 0.5, animations: {
                self.view.layoutIfNeeded()
                self.routinePatternView.alpha = 0.0
            })
        }
    }
    
    @IBAction func routineWeekDayButtonTapped(_ sender: UIButton) {
        self.view.endEditing(true)
        sender.isSelected = !sender.isSelected
        
        if sender.isSelected {
            weekDays.append(String(sender.tag))
        } else if let index = weekDays.index(of: String(sender.tag)) {
            weekDays.remove(at: index)
        }
    }
    
    @IBAction func routineDurationButtonTapped(_ sender: UIButton) {
        self.view.endEditing(true)
        routineDuration2MonthsButton.isSelected = false
        routineDuration3MonthsButton.isSelected = false
        routineDuration4MonthsButton.isSelected = false
        sender.isSelected = true
        routineDurationMonths = sender.tag
    }
    
    @IBAction func directionChanged(_ sender: UISegmentedControl) {
        view.endEditing(true)
        guard let lastOfferedRide = userDefaults.dictionary(forKey: CaronaePreferenceLastOfferedRide.key) else {
            selectedHub = ""
            return
        }
        
        if sender.selectedSegmentIndex == 0, let hubGoing = lastOfferedRide[CaronaePreferenceLastOfferedRide.hubGoing] as? String {
            selectedHub = hubGoing
        } else if sender.selectedSegmentIndex == 1, let hubReturning = lastOfferedRide[CaronaePreferenceLastOfferedRide.hubReturning] as? String {
            selectedHub = hubReturning
        } else {
            selectedHub = ""
        }
    }
    
    @IBAction func selectDateTapped(_ sender: Any) {
        self.view.endEditing(true)
        let title = segmentedControl.selectedSegmentIndex == 0 ? PlaceService.Institution.goingLabel : PlaceService.Institution.leavingLabel
        let datePicker = ActionSheetDatePicker(title: title, datePickerMode: .dateAndTime, selectedDate: self.selectedDate, target: self, action: #selector(timeWasSelected(selectedTime:)), origin: sender)
        datePicker?.minimumDate = Date.currentTimePlus(minutes: 5)
        datePicker?.show()
    }
    
    @IBAction func selectCenterTapped(_ sender: Any) {
        let hubType: HubSelectionViewController.HubTypeDirection = (segmentedControl.selectedSegmentIndex == 0) ? .centers : .hubs
        let selectionVC = HubSelectionViewController(selectionType: .oneSelection, hubTypeDirection: hubType)
        selectionVC.delegate = self
        self.navigationController?.show(selectionVC, sender: self)
    }
    
    @IBAction func selectNeighborhoodTapped(_ sender: Any) {
        let selectionVC = NeighborhoodSelectionViewController(selectionType: .oneSelection)
        selectionVC.delegate = self
        self.navigationController?.show(selectionVC, sender: self)
    }
    
    
    // MARK: Selection Methods
    
    @objc func timeWasSelected(selectedTime: Date) {
        selectedDate = selectedTime
    }
    
    func hasSelected(hubs: [String], inCampus campus: String) {
        self.selectedHub = hubs.first!
    }
    
    func hasSelected(neighborhoods: [String], inZone zone: String) {
        self.selectedZone = zone
        self.selectedNeighborhood = neighborhoods.first!
    }

}


// MARK: UITextViewDelegate

extension CreateRideViewController: UITextViewDelegate {
    func textViewDidBeginEditing(_ textView: UITextView) {
        if textView.text == notesPlaceholder {
            textView.text = ""
            textView.textColor = .black
        }
        textView.becomeFirstResponder()
    }
    
    func textViewDidEndEditing(_ textView: UITextView) {
        if textView.text == "" {
            textView.text = notesPlaceholder
            textView.textColor = notesTextColor
        }
        textView.resignFirstResponder()
    }
}
